src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java

```diff
@@ -639,30 +632,30 @@ default ProgrammingExercise findByIdWithTemplateAndSolutionParticipationTeamAssi
-     * Finds a programming exercise by its id including the submissions for its solution and template participation.
+     * Finds a {@link ProgrammingExercise} by its ID with optional dynamic fetching of associated entities.
      *
-     * @param exerciseId          The id of the programming exercise.
-     * @param withGradingCriteria True, if the grading instructions of the exercise should be included as well.
-     * @return A programming exercise that has the given id.
-     * @throws EntityNotFoundException In case no exercise with the given id exists.
+     * @param exerciseId   the ID of the programming exercise to find.
+     * @param fetchOptions a collection of {@link ProgrammingExerciseFetchOptions} indicating which associated entities to fetch.
+     * @return the {@link ProgrammingExercise} with the specified ID and the associated entities fetched according to the provided options.
+     * @throws EntityNotFoundException if the programming exercise with the specified ID does not exist.
      */
     @NotNull
-    default ProgrammingExercise findByIdWithAuxiliaryRepositoriesTeamAssignmentConfigAndGradingCriteriaElseThrow(long exerciseId, boolean withGradingCriteria)
-            throws EntityNotFoundException {
-
-        final Specification<ProgrammingExercise> specification = (root, query, criteriaBuilder) -> {
-            root.fetch(Exercise_.CATEGORIES, JoinType.LEFT);
-            root.fetch(Exercise_.TEAM_ASSIGNMENT_CONFIG, JoinType.LEFT);
-            root.fetch(ProgrammingExercise_.AUXILIARY_REPOSITORIES, JoinType.LEFT);
-
-            if (withGradingCriteria) {
-                root.fetch(Exercise_.GRADING_CRITERIA, JoinType.LEFT);
-            }
-
-            return null;
-        };
+    default ProgrammingExercise findByIdWithDynamicFetchElseThrow(long exerciseId, ProgrammingExerciseFetchOptions... fetchOptions) throws EntityNotFoundException {
+        return findByIdWithDynamicFetchElseThrow(exerciseId, Set.of(fetchOptions));
+    }
 
-        return findOneByIdElseThrow(specification, exerciseId);
+    /**
+     * Finds a {@link ProgrammingExercise} by its ID with optional dynamic fetching of associated entities.
+     *
+     * @param exerciseId   the ID of the programming exercise to find.
+     * @param fetchOptions a collection of {@link ProgrammingExerciseFetchOptions} indicating which associated entities to fetch.
+     * @return the {@link ProgrammingExercise} with the specified ID and the associated entities fetched according to the provided options.
+     * @throws EntityNotFoundException if the programming exercise with the specified ID does not exist.
+     */
+    @NotNull
+    default ProgrammingExercise findByIdWithDynamicFetchElseThrow(long exerciseId, Collection<ProgrammingExerciseFetchOptions> fetchOptions) throws EntityNotFoundException {
+        var specification = getDynamicSpecification(fetchOptions);
+        return findOneByIdElseThrow(specification, exerciseId, "Programming Exercise");
```

Comment:
Do we actually need to duplicate this method in all repositories that will inherit from `DynamicSpecificationRepository`? We could also add another generic for the FetchOption and implement this method in a generic way

Reply:
Sadly in order to pass the "Programming Exercise" we need to duplicate it. I previously solved this by using the `RepositoryImpl` class, but that was removed a a couple days ago, so I'll have to discuss this with Stephan first.

Using `RepositoryImpl` properly could reduce a lot of code duplication (e.g. all instances of `findByIdElseThrow()`), so I'd address this together in a separate PR

Reply:
I talked with Stephan and will do this in a follow-up where I reintroduce `RepositoryImpl`